File: rotated_sorted.cpp
#include <iostream>
using namespace std;

int search(int A[], int target, int low, int high){

    //if (low > high) return -1;

    while(low <= high){
        int mid = low + (high - low)/2;
        if(target == A[mid]) return mid;   //target  = mid , exit
        else if(A[low] <= A[mid]){         // left side is sorted
            if(A[low] <= target && A[mid] >= target){
            high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        else{
            if(A[mid] <= target && A[high] >= target){  //right side is sorted
                low = mid + 1;
            }
            else{
                high = mid - 1;
            }
        }
    }
    return -1;
}


int main() {
    int A[] = {100, 130, 140, 160, 170, 3, 5, 7, 9, 10, 90};
    int target = 10;
    int n = sizeof(A) / sizeof(A[0]);  //size of the array
    int i = search(A, target, 0, n-1 );

    if( i != -1)
    cout << "Index = " <<i <<endl;
    else
    cout << "target not found" << endl;

}

File: rotatedSorted.py
# Time Complexity: O(logn): This is because each time half of the search space is reduced. We are using Binary Search for it
# Space Complexity: O(1)

# We are using the principle that in a rotated sorted array, atleast one side is sorted.

# Then we first check if the Left is sorted. If the Left side is sorted and the target lies in the range of Low and Mid, we decrease High and do binary search in that half
# Otherwise, we can directly go the the right side

# In a similar fashion we check if the right side is sorted and do the same operation.


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        low = 0
        high = len(nums) - 1

        while low <= high:
            mid = low + (high - low) // 2

            if nums[mid] == target:
                return mid
            elif nums[low] <= nums[mid]:
                # leftissorted
                if nums[low] <= target and nums[mid] > target:
                    high = mid - 1
                else:
                    low = mid + 1
            else:
                if nums[mid] < target and nums[high] >= target:
                    low = mid + 1
                else:
                    high = mid - 1

        return -1
